### PR TITLE
[#960] DigitalHub Migration Export In Batches

### LINKED ISSUE #960: DigitalHub Migration Export In Batches
Change the digitalhub migration export to be done in batches in order to improve performance.

### PR DESCRIPTION
**Pull request by Meowcenary**
Committing memoized header and user lookups, adding extra logging to
help with ongoing debugging, using batches to improve performance of
export. closes #960

### CODE REVIEW

**Comment by Meowcenary** [comment]
Not sure why `subjects_mesh.yml` and `subjects_lcsh.yml` are empty. I think they might have been vestigial.

File: lib/scripts/repo_export_in_batches.rb [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+  converted_record_object = record_converter[:converter_class].new(gf)
+
+  puts "---------\nFile has id: #{gf.id}\n---------"
+  file_path = "tmp/export/#{record_converter[:model_class].name.underscore}_#{gf.id}.json"
+  File.write(file_path, converted_record_object.to_json(pretty: true))
+
+  # increment and put object out of scope
+  generic_file_conversion_count += 1
+  converted_record_object = nil
+end
```
Comment (sharpattack) [inline lib/scripts/repo_export_in_batches.rb:30]: Need to do the same for `Page` model, from line 19-30.

File: lib/scripts/repo_export_in_batches.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Before you run this script empty the directory tmp/export/
+#
+# Run this script with the following:
+#   ./bin/rails r lib/scripts/repo_export_in_batches > "$(date +'%Y-%m-%d-%H%M%S')_repo_export_custom.log"
+#
+puts "---------\nBeginning repo export at #{Time.now} #{Time.zone}\n---------"
+
+# set classes to use for export
+record_converter = {model_class: GenericFile, converter_class: InvenioRdmRecordConverter}
```
Comment (sharpattack) [inline lib/scripts/repo_export_in_batches.rb:9]: Need a `page_converter` for `Page` probably.

File: lib/scripts/repo_export_in_batches.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+  # increment and put object out of scope
+  generic_file_conversion_count += 1
+  converted_record_object = nil
+end
+puts "---------\nCompleted file export at #{Time.now} #{Time.zone}\n---------"
+
+# TODO: Implement collection conversion
+collection_conversion_count = 0
+# Collection.all.each do |collection|
```
Comment (sharpattack) [inline lib/scripts/repo_export_in_batches.rb:35]:
```suggestion
# Collection.find_each do |collection|
```
If we do export `Collection`

File: lib/scripts/repo_export_in_batches.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+GenericFile.find_each do |gf|
+  converted_record_object = record_converter[:converter_class].new(gf)
+
+  puts "---------\nFile has id: #{gf.id}\n---------"
+  file_path = "tmp/export/#{record_converter[:model_class].name.underscore}_#{gf.id}.json"
+  File.write(file_path, converted_record_object.to_json(pretty: true))
+
+  # increment and put object out of scope
+  generic_file_conversion_count += 1
+  converted_record_object = nil
```
Comment (fenekku) [inline lib/scripts/repo_export_in_batches.rb:29]: Just curious about this: why is that necessary? Isn't the line https://github.com/galterlibrary/digital-repository/pull/961/files#diff-ec7071bf61726d0fda71d83e812c135e96ef7c7fc6ce7ad9bf5460a5e43417bbR21 above getting rid of the only reference to the "converted record object"?

Comment (Meowcenary) [reply lib/scripts/repo_export_in_batches.rb]: Yeah, this was done in error. I read some post about triggering the garbage collector by removing all references to objects, but then I applied it here where it doesn't really fit.

File: lib/scripts/repo_export_in_batches.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
+# Before you run this script empty the directory tmp/export/!
+#   rm -rf tmp/export/; mkdir tmp/export/
```
Comment (sharpattack) [inline lib/scripts/repo_export_in_batches.rb:2]: Do we need to empty the directory? I may be reluctant to knowing how long it has been taking. Takes a while to create the exported json files, and one second to remove them all 😨

**Review by sharpattack — APPROVED** [review]
Looks good to me, thanks for refactoring/improving. Let's get this released and on production.

File: lib/scripts/repo_export_in_batches.rb (hunk 5 above)
Comment (Meowcenary) [reply lib/scripts/repo_export_in_batches.rb:2]: Good point. I guess we can copy the contents out, then remove it.

**Review by carsonicator — COMMENTED** [review]
Thanks @Meowcenary

**Comment by sharpattack** [comment]
There has been a mix up here. Somehow, the commit(or commit hash) in this PR did not make it into this release PR #962 So `production` branch now has a "rogue" commit(1a21f3b0ba2b9d0d14ca004133d04c8dc0583e5a). To resolve, I will need to delete the `release` branch, and drop the "rogue" commit from `production` and force push it. 

Screen Shot 2021-12-30 at 4 07 25 PM
Screen Shot 2021-12-30 at 4 11 45 PM
Screen Shot 2021-12-30 at 4 12 27 PM
